### ISSUE TITLE
PhaseTwo_RollbackFailed_Retryable because of JSON deserialize lost type reference

### ISSUE THREAD

**Issue by helloworlde** [seata/seata#1188]
- [x] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

There has same title issue #1121 , but looks like it's not fixed this issue.

### Ⅰ. Issue Description

I used Seata for Spring Cloud, and Seata version is 0.6.1

When I get an exception and it should be rollback, but it's always failed and status is `PhaseTwo_RollbackFailed_Retryable`. After change Seata version as 0.5.2, it work correct.

And it caused by JSON deserialized lost type reference.

### Ⅱ. Describe what happened

I found there called validate method in [io.seata.rm.datasource.undo.AbstractUndoExecutor#executeOn](https://github.com/seata/seata/blob/develop/rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java#L102), when compare beforeRecords and currentRecords for field `pay_amount` in [io.seata.rm.datasource.DataCompareUtils#isFieldEquals](https://github.com/seata/seata/blob/develop/rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java#L48), it return false. And I found the beforeRecords's filed `pay_amount` class type is Integer, but the currentRecords is BigDecimal. And the column is decimal too. 

Finally, I found the difference come from [io.seata.rm.datasource.undo.JSONBasedUndoLogParser#decode](https://github.com/seata/seata/blob/develop/rm-datasource/src/main/java/io/seata/rm/datasource/undo/parser/FastjsonUndoLogParser.java#L53). The BigDecimal was converted to Integer after deserialized. 

### Ⅲ. Describe what you expected to happen

Transaction should be rollbacked success.

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

The simple way to reproduce code is :

```java
public class FastJsonUsage {

    public static void main(String[] args) {

        List<Field> fields = new ArrayList<>();
        fields.add(Field.builder().name("Integer").type(3).value(1).build());
        fields.add(Field.builder().name("BigDecimal").type(3).value(BigDecimal.ONE).build());
        System.out.println("Before");
        fields.forEach(f -> System.out.println(f.getName() + ":" + f.getValue().getClass().getName()));

        String serializedString = JSON.toJSONString(fields, SerializerFeature.WriteDateUseDateFormat);

        List<Field> deserializeList = JSON.parseArray(serializedString, Field.class);

        System.out.println("After");
        deserializeList.forEach(f -> System.out.println(f.getName() + ":" + f.getValue().getClass().getName()));

    }

    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    @Getter
    @Setter
    static class Field {
        private String name;
        private Integer type;
        private Object value;
    }
}

```

and the result is 

```
Before
Integer:java.lang.Integer
BigDecimal:java.math.BigDecimal
After
Integer:java.lang.Integer
BigDecimal:java.lang.Integer
```

You can reproduce with my project also, Which under module [cloud-seata-nacos](https://github.com/helloworlde/spring-cloud-alibaba-component/tree/master/cloud-seata-nacos)

### Ⅴ. Anything else we need to know?

I suggest to add class type correct action after decode for Number or other class which may  lost type reference since you had recorded type in `rollback_info`

**Comment by helloworlde** [comment]
Finally I found it had been fixed is PR #1125 by using Protobuf as serialization, but the PR #1132 still can't cover this issue.

**Comment by slievrly** [comment]
@helloworlde Are you sure you can't meet expectations with jackson? Please paste the file.conf and the create table script

**Comment by helloworlde** [comment]
No, I mean the exception won't occur only by using Protobuf as serialization. Neither fastjson nor jackson work correct. So maybe Protobuf should be the default configuration.

You can reference my project [spring-cloud-alibaba-component/cloud-seata-nacos](https://github.com/helloworlde/spring-cloud-alibaba-component/tree/master/cloud-seata-nacos)

I used Nacos save configuration. 

SQL script is :
```sql
DROP DATABASE IF EXISTS seata;
CREATE DATABASE seata;
USE seata;

CREATE TABLE account
(
    id               INT AUTO_INCREMENT PRIMARY KEY,
    balance          DECIMAL,
    last_update_time DATETIME DEFAULT now() ON UPDATE now()
);

CREATE TABLE product
(
    id               INT AUTO_INCREMENT PRIMARY KEY,
    price            DECIMAL,
    stock            INT,
    last_update_time DATETIME DEFAULT now() ON UPDATE now()
);

CREATE TABLE orders
(
    id               INT AUTO_INCREMENT PRIMARY KEY,
    user_id          INT,
    product_id       INT,
    pay_amount       DECIMAL,
    status           VARCHAR(100),
    add_time         DATETIME DEFAULT now(),
    last_update_time DATETIME DEFAULT now() ON UPDATE now()
);

INSERT INTO account (id, balance)
VALUES (1, 100);
INSERT INTO product (id, price, stock)
VALUES (1, 5.00, 10);

CREATE TABLE undo_log
(
    id            BIGINT(20)   NOT NULL AUTO_INCREMENT,
    branch_id     BIGINT(20)   NOT NULL,
    xid           VARCHAR(100) NOT NULL,
    rollback_info LONGBLOB     NOT NULL,
    log_status    INT(11)      NOT NULL,
    log_created   DATETIME     NOT NULL,
    log_modified  DATETIME     NOT NULL,
    ext           VARCHAR(100) DEFAULT NULL,
    PRIMARY KEY (id),
    UNIQUE KEY ux_undo_log (xid, branch_id)
) ENGINE = InnoDB
  AUTO_INCREMENT = 1
  DEFAULT CHARSET = utf8;
```

And `nacos-config.txt` is:

```
transport.type=TCP
transport.server=NIO
transport.heartbeat=true
transport.thread-factory.boss-thread-prefix=NettyBoss
transport.thread-factory.worker-thread-prefix=NettyServerNIOWorker
transport.thread-factory.server-executor-thread-prefix=NettyServerBizHandler
transport.thread-factory.share-boss-worker=false
transport.thread-factory.client-selector-thread-prefix=NettyClientSelector
transport.thread-factory.client-selector-thread-size=1
transport.thread-factory.client-worker-thread-prefix=NettyClientWorkerThread
transport.thread-factory.boss-thread-size=1
transport.thread-factory.worker-thread-size=8
service.vgroup_mapping.order-service-fescar-service-group=default
service.vgroup_mapping.pay-service-fescar-service-group=default
service.vgroup_mapping.storage-service-fescar-service-group=default
service.enableDegrade=false
service.disable=false
service.max.commit.retry.timeout=-1
service.max.rollback.retry.timeout=-1
client.async.commit.buffer.limit=10000
client.lock.retry.internal=10
client.lock.retry.times=30
store.mode=file
store.file.dir=file_store/data
store.file.max-branch-session-size=16384
store.file.max-global-session-size=512
store.file.file-write-buffer-cache-size=16384
store.file.flush-disk-mode=async
store.file.session.reload.read_size=100
store.db.datasource=dbcp
store.db.db-type=mysql
store.db.url=jdbc:mysql://127.0.0.1:3306/seata?useUnicode=true
store.db.user=mysql
store.db.password=mysql
store.db.min-conn=1
store.db.max-conn=3
store.db.global.table=global_table
store.db.branch.table=branch_table
store.db.query-limit=100
lock.mode=memory
lock.db.lock-table=lock_table
recovery.committing-retry-delay=5
recovery.asyn-committing-retry-delay=5
recovery.rollbacking-retry-delay=5
recovery.timeout-retry-delay=5
transaction.undo.data.validation=true
```

**Comment by cherrylzhao** [comment]
I have also found this bug when I integrated with ShardingSphere.
column type from undo log is incorrect, so rollback execute failed due to data consistence check.

````
afterRecords:
  order_item_id: Integer (5)

currentRecords:
  order_item_Id: Long  (5)

```

order_item_id should be Long type

**Comment by jsbxyyx** [comment]
I use jackson test you provide sql, no problems found.

**Comment by helloworlde** [comment]
Had fixed in release 0.7.0
